### PR TITLE
Fixes generic inference in functions with `TypeGuard`, refs #11780

### LINKED ISSUE #11428: TypeGuard functions not properly recognizing generic types
**Bug Report**

It seems the `typing.TypeGuard` support is not able to properly discriminate a union of two (generic) types.

**To Reproduce**

* The example below defines two completely distinct types, `Left[L]` and `Right[R]`, both of which happen to be generic and have a `.value` attribute.
* The `LeftOrRight` is a `Union` of those types.
* `x` is a variable of type `LeftOrRight[str, int]`. In this example it has value `Left("")`
* Using `isinstance()` works fine to figure out whether `x` is either `Left` or `Right`, and within the `if/else` block the types are correct, including the types `L` and `R` (both `TypeVar`)
* The `is_left()` and `is_right()` helpers use `typing.TypeGuard`, but Mypy does not recognize the types like it did with the `isinstance()` check.

```python
from typing import Generic, TypeGuard, TypeVar, Union

L = TypeVar("L")
R = TypeVar("R")


class Left(Generic[L]):
    value: L

    def __init__(self, value: L):
        self.value = value


class Right(Generic[R]):
    value: R

    def __init__(self, value: R):
        self.value = value


LeftOrRight = Union[Left[L], Right[R]]


def is_left(obj: LeftOrRight[L, R]) -> TypeGuard[Left[L]]:
    return isinstance(obj, Left)


def is_right(obj: LeftOrRight[L, R]) -> TypeGuard[Right[R]]:
    return isinstance(obj, Right)


x: LeftOrRight[str, int] = Left("")
reveal_type(x)  # Revealed type is "Union[Left[builtins.str], Right[builtins.int]]"

if isinstance(x, Left):
    reveal_type(x)  # Revealed type is "Left[builtins.str]"
    x.value + ""  # only valid for str
else:
    reveal_type(x)  # Revealed type is "Right[builtins.int]"
    x.value + 10  # only valid for int

if is_left(x):
    reveal_type(x)  # Revealed type is "Left[L`-1]"
    x.value + ""  #  Unsupported left operand type for + ("L")  [operator]
else:
    reveal_type(x)  # Revealed type is "Union[Left[builtins.str], Right[builtins.int]]"
    x.value + 10  # Unsupported operand types for + ("str" and "int")  [operator]

```

**Expected Behavior**

`is_left()` and `is_right()` act as type-safe ways to discriminate between the two members of the `Union`, including the actual type of the (generic) `.value` attribute.

**Actual Behavior**

* `reveal_type()` returns wrong results
* type errors reported in what seems to be perfectly fine and very strict code

**Your Environment**

This happens with both Mypy 0.910 (current release as time of writing) and today's `git` checkout of the main branch.

### LINKED ISSUE #11780: `TypeGuard` doesn't work with `TypeVar`
**Bug Report**

Using a `TypeVar` with `TypeGuard` results in Mypy thinking that the resulting type is the `TypeVar` itself, and not the type it should be pointing to.

**To Reproduce**

Example adapted from [PEP 647](https://www.python.org/dev/peps/pep-0647/#typeguard-type):
```py
# src/test.py

from typing import TypeGuard, TypeVar

_T = TypeVar("_T")


def is_two_element_tuple(val: tuple[_T, ...]) -> TypeGuard[tuple[_T, _T]]:
    return len(val) == 2


def func(names: tuple[str, ...]):
    if is_two_element_tuple(names):
        reveal_type(names)  # Tuple[str, str]
        for name in names:
            print(name.capitalize(), "- two")
    else:
        reveal_type(names)  # Tuple[str, ...]
        for name in names:
            print(name.capitalize(), "- not two")

```

**Expected Behavior**

Mypy should substitute the `TypeVar` inside `TypeGuard` to the proper type it should be pointing to.
Pyre, Pyright, and Pylance do successfully work with the `TypeVar` inside `TypeGuard`.

```
$ pyre
ƛ No type errors found
src/test.py:14:8 Revealed type [-1]: Revealed type for `names` is `typing.Tuple[str, str]`.
src/test.py:18:8 Revealed type [-1]: Revealed type for `names` is `typing.Tuple[str, ...]`.

$ pyright src/test.py
/home/micael/projects/python-3_10/src/test.py
  /home/micael/projects/python-3_10/src/test.py:14:21 - info: Type of "names" is "tuple[str, str]"
  /home/micael/projects/python-3_10/src/test.py:18:21 - info: Type of "names" is "tuple[str, ...]"
0 errors, 0 warnings, 2 infos
```

**Actual Behavior**

```
$ mypy src/test.py
src/test.py:14: note: Revealed type is "Tuple[_T`-1, _T`-1]"
src/test.py:16: error: "_T" has no attribute "capitalize"
src/test.py:18: note: Revealed type is "builtins.tuple[builtins.str]"
src/test.py:19: error: Incompatible types in assignment (expression has type "str", variable has type "_T")
src/test.py:20: error: "_T" has no attribute "capitalize"
Found 3 errors in 1 file (checked 1 source file)
```

**Your Environment**

| Name | Version |
| -- | -- |
| Python | `3.10.0` |
| typing_extensions | `4.0.1` |
| mypy | `0.920` |
| mypy-extensions | `0.4.3` |
| pyre-check | `0.9.8` |
| pyre-extensions | `0.0.23` |
| pyright | `0.0.13` |

- Mypy version used: `0.920`
- Mypy command-line flags: none
- Mypy configuration options from `mypy.ini` (and other config files): none
- Python version used: `3.10.0`
- Operating system and version: Manjaro KDE

> Operating System: Manjaro Linux
KDE Plasma Version: 5.23.4
KDE Frameworks Version: 5.88.0
Qt Version: 5.15.2
Kernel Version: 5.15.7-1-MANJARO (64-bit)
Graphics Platform: X11
Processors: 4 × AMD Ryzen 3 2200G with Radeon Vega Graphics
Memory: 7.7 GiB of RAM
Graphics Processor: NVIDIA GeForce GTX 970/PCIe/SSE2

### PR DESCRIPTION
**Pull request by sobolevn**
Closes #11780, closes #11428

### CODE REVIEW

**Comment by intgr** [comment]
Thanks!

I have a similar use case with literals. I think this change helps a little, but my example runs into a different issue: `Literal[...]` expression is not considered a type in the type system:

```python
from typing import Literal, TypeGuard, TypeVar, get_args

T = TypeVar('T')

def is_literal_value(value: str | int | None, literal_type: type[T]) -> TypeGuard[T]:
    return value in get_args(literal_type)

value: str = input()
if is_literal_value(value, Literal['on', 'off']):
    reveal_type(value)
```

Expected result: `Revealed type is Literal['on', 'off']`

Output in mypy master 😕
```
_local/typeguard.py:8: error: Argument 2 to "is_literal_value" has incompatible type "object"; expected "Type[<nothing>]"
_local/typeguard.py:9: note: Revealed type is "T`-1"
```

Output from this branch
```
_local/typeguard.py:8: error: Argument 2 to "is_literal_value" has incompatible type "object"; expected "Type[<nothing>]"
_local/typeguard.py:9: note: Revealed type is "<nothing>"
```

**Comment by JelleZijlstra** [comment]
Obligatory reference to #9773.

**Comment by intgr** [comment]
Thanks @JelleZijlstra, I have been searching for discussion on this topic, but had not found that issue.

**Comment by goodoldneon** [comment]
Running into this problem, too.

I get these errors:

```
main.py:23: note: Revealed type is "builtins.list[Any]"
main.py:26: note: Revealed type is "builtins.list[T`-1]"
main.py:27: error: Argument 1 to "do_stuff" has incompatible type "List[T]"; expected "List[Foo]"
Found 1 error in 1 file (checked 1 source file)
```

When I run this code:

```py
from typing import Any, TypeGuard, Type, TypeVar

def do_stuff(foos: list[Foo]) -> None:
    pass

class Foo:
    pass

T = TypeVar("T")

def is_list_of(value: list, type: Type[T]) -> TypeGuard[list[T]]:
    for item in value:
        if isinstance(item, type) is False:
            return False

    return True

foos: list[Any] = [Foo()]
reveal_type(foos)

if is_list_of(foos, Foo):
    reveal_type(foos)
    do_stuff(foos)
```

**Comment by sobolevn** [comment]
I think we can merge this. CC @JukkaL

**Review by hauntsaninja — APPROVED** [review]
This is great!

I think this might change behaviour if you have a function called `_` that is type guarded, but that's fine.

File: test-data/unit/check-typeguard.test [hunk 1]
```diff
@@ -547,3 +547,53 @@ accepts_typeguard(with_typeguard_a)  # E: Argument 1 to "accepts_typeguard" has
...
+
+def func1(name: _T):
+    reveal_type(name)  # N: Revealed type is "_T`-1"
+    if identity(name):
+        reveal_type(name)  # N: Revealed type is "_T`-1"
+
+def func2(name: str):
+    reveal_type(name)  # N: Revealed type is "builtins.str"
+    if identity(name):
+        reveal_type(name)  # N: Revealed type is "builtins.str*"
```
Comment (hauntsaninja) [inline test-data/unit/check-typeguard.test:568]:
```suggestion
        reveal_type(name)  # N: Revealed type is "builtins.str"
```

File: test-data/unit/check-typeguard.test [hunk 2]
```diff
@@ -547,3 +547,53 @@ accepts_typeguard(with_typeguard_a)  # E: Argument 1 to "accepts_typeguard" has
...
+
+_T = TypeVar("_T")
+
+def is_list_of_str(val: _T) -> TypeGuard[List[_T]]:
+    pass
+
+def func(name: str):
+    reveal_type(name)  # N: Revealed type is "builtins.str"
+    if is_list_of_str(name):
+        reveal_type(name)  # N: Revealed type is "builtins.list[builtins.str*]"
```
Comment (hauntsaninja) [inline test-data/unit/check-typeguard.test:583]:
```suggestion
        reveal_type(name)  # N: Revealed type is "builtins.list[builtins.str]"
```

File: test-data/unit/check-typeguard.test [hunk 3]
```diff
@@ -547,3 +547,53 @@ accepts_typeguard(with_typeguard_a)  # E: Argument 1 to "accepts_typeguard" has
...
+
+_T = TypeVar("_T")
+
+def is_two_element_tuple(val: Tuple[_T, ...]) -> TypeGuard[Tuple[_T, _T]]:
+    pass
+
+def func(names: Tuple[str, ...]):
+    reveal_type(names)  # N: Revealed type is "builtins.tuple[builtins.str]"
+    if is_two_element_tuple(names):
+        reveal_type(names)  # N: Revealed type is "Tuple[builtins.str*, builtins.str*]"
```
Comment (hauntsaninja) [inline test-data/unit/check-typeguard.test:598]:
```suggestion
        reveal_type(names)  # N: Revealed type is "Tuple[builtins.str, builtins.str]"
```

File: test-data/unit/check-typeguard.test [hunk 4]
```diff
@@ -547,3 +547,53 @@ accepts_typeguard(with_typeguard_a)  # E: Argument 1 to "accepts_typeguard" has
...
+from typing import TypeVar, Tuple
+from typing_extensions import TypeGuard
+
+_T = TypeVar("_T")
+
+def is_two_element_tuple(val: Tuple[_T, ...]) -> TypeGuard[Tuple[_T, _T]]:
+    pass
+
+def func(names: Tuple[str, ...]):
+    reveal_type(names)  # N: Revealed type is "builtins.tuple[builtins.str]"
```
Comment (hauntsaninja) [inline test-data/unit/check-typeguard.test:596]:
```suggestion
    reveal_type(names)  # N: Revealed type is "builtins.tuple[builtins.str, ...]"
```

**Pull request merged**